### PR TITLE
Add option to nest context within the h1 for title component

### PR DESCRIPTION
**Pull request by jon-kirwan**
## What
Move the caption inside of the h1.

## Why
Currently, the context is missed when page titles are read by screen readers. This way when users navigate through headings with a screenreader will hear the title with context announced.

No visual changes.

https://trello.com/c/gPFqiwZa/799-update-headings-for-smart-answers-outcome

### CODE REVIEW

File: app/views/govuk_publishing_components/components/_title.html.erb [hunk 1]
```diff
@@ -22,12 +22,12 @@
...
 <%= content_tag(:div, class: classes) do %>
-  <% if context %>
+  <h1 class="<%= heading_classes.join(" ") %>">
+    <% if context %>
     <span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
       <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
     </span>
-  <% end %>
-  <h1 class="<%= heading_classes.join(" ") %>">
+    <% end %>
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_title.html.erb:30]: Nitpick: the span inside this `if` statement should be indented.

**Review by andysellick — APPROVED** [review]
Looks good. One small thing - the PR description doesn't explicitly state which component this change refers to, I had to figure it out from the files changed.

**Review by maxgds — COMMENTED** [review]
I feel like this ability to put the context within the H1 should be an option rather than the default. The design system says "If the caption should be considered part of the page heading, you can also nest the caption within the h1." and I'm not sure that we want to retroactively update all instances of it that are on the site already - but I'm not sure that we don't, either.  I'm going to open a discussion on the govuk frontenders Slack channel.

**Comment by jon-kirwan** [comment]
Thanks all. @andysellick. I've updated the PR to indicate the title component has been updated and have amended indentation in the template too.

@maxgds - i've now added this as an option instead so it can be implemented like so -

```ruby
<%= render "govuk_publishing_components/components/title", {
  context: "Publication",
  title: "My page title",
  context_inside: true
} %>
```

I've also updated the docs and tests.

I'm still quite new to Ruby so would appreciate any feedback on a better way to reference the context (which is currently repeated for both context options) -

```ruby
<span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
  <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
</span>
```

**Comment by maxgds** [comment]
Looking good. I can see there are going to be some conflicts with the code I updated yesterday which removes the context href, so you're probably going to need a rebase against master and might have to untangle that a bit, sorry.

In terms of the repetitive context, you could use a content_for block to define it once in the template, see docs here:
https://apidock.com/rails/ActionView/Helpers/CaptureHelper/content_for

I'd also suggest a tweak to your tests - in each one checking for context I'd add a negative assertion for the other position - so we know it just appears once. Docs for that are here: https://apidock.com/rails/ActionController/Assertions/SelectorAssertions/assert_select

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@
   useful summary for people upgrading their application, not a replication
   of the commit log.
 
+## Unreleased
+
+* Add option to nest the context within the `<h1>` ([PR #2226](https://github.com/alphagov/govuk_publishing_components/pull/2226))
```
Comment (andysellick) [inline CHANGELOG.md:12]: This should match the PR title.

File: Gemfile.lock [hunk 3]
```diff
@@ -74,7 +74,7 @@ GEM
     addressable (2.8.0)
       public_suffix (>= 2.0.2, < 5.0)
     ast (2.4.2)
-    brakeman (5.0.4)
+    brakeman (5.1.1)
```
Comment (andysellick) [inline Gemfile.lock:77]: Changes to dependencies in `Gemfile.lock` shouldn't be included in a PR to change a component unless those dependencies are related to the change - which in this case I think they're not. Can these changes be removed?

File: app/views/govuk_publishing_components/components/_title.html.erb [hunk 4]
```diff
@@ -21,13 +22,20 @@
...
+  <% if context && !context_inside %>
     <span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
       <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
     </span>
   <% end %>
+  
   <h1 class="<%= heading_classes.join(" ") %>">
+    <% if context && context_inside %>
+      <span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
+        <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_title.html.erb:36]: It looks like these lines are the same as lines 28-30? In which case you can use the rails `capture` method, something like this:

```
<% context_block = capture do %>
    <span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
      <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
    </span>
<% end %>

...

<% if context && !context_inside %>
  <%= context_block %> 
<% end %>

<h1>
  <% if context &&  context_inside %>
    <%= context_block %>
  <% end %>
</h1>

```

File: app/views/govuk_publishing_components/components/docs/title.yml [hunk 5]
```diff
@@ -24,6 +24,13 @@ examples:
     data:
       context: Publication
       title: My page title
+  with_context_inside:
+    description: |
+      If the context should be considered part of the page heading, you can also nest the context within the <code><h1></code>.
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/title.yml:29]: Nitpick: 'also' is unnecessary, `...you can nest the context within...`

File: spec/components/title_spec.rb [hunk 6]
```diff
@@ -21,6 +21,11 @@ def component_name
     assert_select ".govuk-caption-xl", text: "Format"
   end
 
+  it "title context inside appears" do
```
Comment (andysellick) [inline spec/components/title_spec.rb:24]: `title context appears inside` maybe?

**Comment by jon-kirwan** [comment]
Thanks both. All changes above have now been added / amended.

**Review by maxgds — APPROVED** [review]
Looks good to me!
